src/bin/pg_autoctl/pgsql.c

```diff
@@ -165,6 +205,115 @@ pgsql_open_connection(PGSQL *pgsql)
 }
 
 
+/*
+ * pgsql_retry_open_connection loops over a PQping call until the remote server
+ * is ready to accept connections, and then connects to it and returns true
+ * when it could connect, false otherwise.
+ */
+static PGconn *
+pgsql_retry_open_connection(PGSQL *pgsql)
+{
+	PGconn *connection = NULL;
+	int attempts = 0;
+	bool retry = true;
+	bool connectionOk = false;
+	uint64_t startTime = time(NULL);
+
+	log_warn("Failed to connect to \"%s\", retrying until "
+			 "the server is ready", pgsql->connectionString);
+
+	while (retry)
+	{
+		uint64_t now = time(NULL);
+
+		if ((now - startTime) >= POSTGRES_PING_RETRY_TIMEOUT)
+		{
+			log_warn("Failed to connect to \"%s\" after %d attempts, "
+					 "stopping now", pgsql->connectionString, attempts);
+			break;
+		}
+
+		++attempts;
+		switch (PQping(pgsql->connectionString))
+		{
+			case PQPING_OK:
+			{
+				log_debug("PQping OK after %d attempts", attempts);
+				retry = false;
+
+				/*
+				 * Ping is now ok, and connection is still NULL because the
+				 * first attempt to connect failed. Now is a good time to
+				 * establish the connection.
+				 *
+				 * PQping does not check authentication, so we might still fail
+				 * to connect to the server.
+				 */
+				connection = PQconnectdb(pgsql->connectionString);
+
+				if (PQstatus(connection) == CONNECTION_OK)
+				{
+					connectionOk = true;
+				}
+				else
+				{
+					log_error("Failed to connect after successful "
+							  "ping, please verify authentication "
+							  "and logs on the server at \"%s\"",
+							  pgsql->connectionString);
+				}
+				break;
+			}
+
+			case PQPING_REJECT:
+			{
+				log_error("Connection rejected: \"%s\"",
+						  pgsql->connectionString);
+				retry = false;
+				break;
+			}
+
+			case PQPING_NO_RESPONSE:
+			{
+				log_debug("PQping: no response after %d attempts",
+						  attempts);
+				retry = true;
+				break;
+			}
+
+			case PQPING_NO_ATTEMPT:
+			{
+				log_error("Failed to ping server \"%s\" because of "
+						  "client-side problems (no attempt were made)",
+						  pgsql->connectionString);
```

Comment:
Would pgsql->connectionstring contain passwords? It might be better to strip passwords out of the logs.